### DISCUSSION TITLE
kv store feature

### DISCUSSION THREAD

**Discussion by carr123** [nats-io/nats-server#2487] [Q&A] [answered]
hello, i learned from nats.io that there are key value and object store on the roadmap.
i don't find detailed description or doc about that cool feature.
is that something like redis or Amazon S3 ?

**Comment by derekcollison** [comment, marked as answer]
The functionality is enabled via MaxMsgsPerSubject setting for a stream.

https://www.youtube.com/watch?v=yYL0ZGTNomE

The NATS cli has support and some clients have it as well already like Node.js and Java IIRC.

**Comment by ripienaar** [comment]
Any cluster can read a KV bucket hosted in other regions of a super cluster. 

They support replication between regions as mentioned in the video and read replicas - though still formalising that. 

The nats.go client for the feature in preview mode this week.

**Comment by ripienaar** [comment]
The ADR is linked below but note we have a PR in progress to bring it up to date with nats.go implementation. 

https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-8.md
